src/physics/grid/ValueGridBuilder.cc

```diff
@@ -162,33 +184,85 @@ ValueGridLogBuilder::ValueGridLogBuilder(real_type              emin,
 /*!
  * Get the storage type and requirements for the energy grid.
  */
-auto ValueGridLogBuilder::energy_storage() const -> EnergyStorage
+auto ValueGridLogBuilder::storage() const -> Storage
 {
-    return {EnergyLookup::uniform_log, 0};
+    return {ValueGridType::xs, xs_.size()};
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Get the storage type and requirements for the value grid.
+ * Construct on device.
  */
-auto ValueGridLogBuilder::value_storage() const -> ValueStorage
+void ValueGridLogBuilder::build(ValueGridStore* store) const
 {
-    return {ValueCalculation::linear, xs_.size()};
+    XsGridPointers ptrs;
+
+    // Construct log(energy) grid
+    ptrs.log_energy
+        = UniformGridPointers::from_bounds(log_emin_, log_emax_, xs_.size());
+
+    // Provide reference to values
+    ptrs.value = make_span(xs_);
+
+    // Copy data to store
+    store->push_back(ptrs);
+}
+
+//---------------------------------------------------------------------------//
+// GENERIC BUILDER
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from raw data.
+ */
+ValueGridGenericBuilder::ValueGridGenericBuilder(VecReal grid,
+                                                 VecReal value,
+                                                 Interp  grid_interp,
+                                                 Interp  value_interp)
+    : grid_(std::move(grid))
+    , value_(std::move(value))
+    , grid_interp_(grid_interp)
+    , value_interp_(value_interp)
+{
+    CELER_EXPECT(grid_.size() >= 2
+                 && is_monotonic_increasing(make_span(grid_)));
+    CELER_EXPECT(value_.size() == grid_.size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from raw data with linear interpolation.
+ */
+ValueGridGenericBuilder::ValueGridGenericBuilder(VecReal grid, VecReal value)
+    : ValueGridGenericBuilder(
+        std::move(grid), std::move(value), Interp::linear, Interp::linear)
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the storage type and requirements for the energy grid.
+ */
+auto ValueGridGenericBuilder::storage() const -> Storage
+{
+    return {ValueGridType::generic, 2 * grid_.size()};
 }
 
 //---------------------------------------------------------------------------//
 /*!
  * Construct on device.
  */
-void ValueGridLogBuilder::build(ValueGridStore*) const
+void ValueGridGenericBuilder::build(ValueGridStore* store) const
```

Comment:
Do we have a decision/documentation on when to pass a pointer and when to pass a reference to a function that always expect a non-nulllptr input (I.e. does not have a `if (store == nullptr) return`?  (For example in this PR we see some function that takes a pointer (without a nullptr check) and some that takes a reference, like ValueGridStore::push_back).

Reply:
Hm, I thought at one point I'd codified "as with the google style guide, use pointers rather than mutable references" to make it more obvious in the calling code what's potentially being modified. Your point is well  taken that an extra assertion at the beginning of the function would make it clear that this isn't an omission.

Reply:
For reference, it was discussed here: https://github.com/celeritas-project/celeritas/pull/55#discussion_r511473298 (though not sure if it's documented anywhere else).

Reply:
Thanks for the reference Amanda. I've cleaned up that comment and added it to the development guidance document https://github.com/celeritas-project/celeritas/wiki/Development#function-arguments-and-return-values .